modele-social/règles/entreprise-établissement.yaml

```diff
@@ -110,25 +110,151 @@ entreprise . imposition . IS . notification IS non intégré:
+
+entreprise . exercice . durée:
+  titre: durée de l'exercice
+  formule:
+    durée:
+      depuis: début
+      jusqu'à: fin
+
+entreprise . exercice . date trop ancienne:
+  type: notification
+  sévérité: avertissement
+  formule: début < 01/01/2018
+  description: La date saisie est trop ancienne. Le simulateurs n'intègre pas les barèmes avant 2018.
+
+entreprise . exercice . date trop éloignée:
+  type: notification
+  sévérité: avertissement
+  formule: fin > 31/12/2022
+  description: La date saisie est trop éloignée. Le simulateurs n'intègre pas les barèmes au delà de 2022.
+
+entreprise . exercice . début après la fin:
+  type: notification
+  sévérité: avertissement
+  formule: début >= fin
+  description: La fin de l'exercice doit être postérieure à celle de son début.
+
+entreprise . exercice . durée maximale:
+  type: notification
+  sévérité: avertissement
+  formule: durée >= 24 mois
+  description: La durée maximale d'un exercice comptable est de 24 mois.
+
 entreprise . impôt sur les sociétés:
   unité: €/an
   formule:
     barème:
       assiette: bénéfice
-      tranches:
-        - taux: 15%
-          plafond: 38120 €/an
-        - taux: 28%
-          plafond: 500000 €/an
-        - taux: 33.3%
+      variations:
+        - si: exercice . début >= 01/01/2022
+          alors:
+            tranches:
+              - taux: 15%
+                plafond: 38120 €/an * prorata taux réduits
+              - taux: 25%
+        - si: exercice . début >= 01/01/2021
+          alors:
+            tranches:
+              - taux: 15%
+                plafond: 38120 €/an * prorata taux réduits
+              - taux: 26.5%
+        - si: exercice . début >= 01/01/2020
+          alors:
+            tranches:
+              - taux: 15%
+                plafond: 38120 €/an * prorata taux réduits
+              - taux: 28%
+        - si: exercice . début >= 01/01/2019
+          alors:
+            tranches:
+              - taux: 15%
+                plafond: 38120 €/an * prorata taux réduits
+              - taux: 28%
+                plafond: 500000 €/an * prorata taux réduits
+              - taux: 31%
+        - si: exercice . début >= 01/01/2018
+          alors:
+            tranches:
+              - taux: 15%
+                plafond: 38120 €/an * prorata taux réduits
+              - taux: 28%
+                plafond: 500000 €/an * prorata taux réduits
+              - taux: 33.3333%
+    arrondi: oui
+  références:
+    Fiche impots.gouv.fr: https://www.impots.gouv.fr/portail/international-professionnel/impot-sur-les-societes
+    Fiche service-public.fr: https://www.service-public.fr/professionnels-entreprises/vosdroits/F23575
+
+entreprise . impôt sur les sociétés . prorata taux réduits:
+  unité: '%'
+  formule:
+    variations: 
+      - si: éligible taux réduits
+        alors: prorata durée de l'exercice
+      - sinon: 0 €/an
+  note: |
+    Publicodes ne nous permet pas de définir qu'une tranche "taux réduit" d'un
```

Comment:
Ah, c'est pas beau ça. Faudrait qu'on arrive à s'en sortir autrement... J'imagine qu'avec la gestion des listes, on pourrait s'en sortir de manière élégante avec 
```
- taux: 15%
  plafond: 2000€
  applicable si: éligible taux réduit
```

En attendant, on a mieux : l'applicabilité peut-être déplacée sur le plafond, et ça marche (même si la visualisation est un peu moche) : 
https://publi.codes/studio/IS?code=%0A%23%20Bienvenue%20dans%20le%20bac%20%C3%A0%20sable%20du%20langage%20publicode%20!%0A%23%20Pour%20en%20savoir%20plus%20sur%20le%20langage%20%3A%0A%23%20%3D%3E%20https%3A%2F%2Fpubli.codes%2Fdocumentation%2Fprincipes-de-base%0A%0Ataux%20r%C3%A9duit%3A%20oui%0AIS%3A%0A%20%20bar%C3%A8me%3A%0A%20%20%20%20assiette%3A%201000%E2%82%AC%2Fan%0A%20%20%20%20tranches%3A%0A%20%20%20%20-%20taux%3A%2015%25%0A%20%20%20%20%20%20plafond%3A%20%0A%20%20%20%20%20%20%20%20valeur%3A%2038120%20%E2%82%AC%2Fan%0A%20%20%20%20%20%20%20%20applicable%20si%3A%20taux%20r%C3%A9duit%0A%20%20%20%20-%20taux%3A%2028%25%0A%20%20%20%20%20%20plafond%3A%20500000%20%E2%82%AC%2Fan%0A%20%20%20%20-%20taux%3A%2031%25

Reply:
Oui j'avais 

```yaml
- taux: 15%
  plafond: 2000€
  applicable si: éligible taux réduit
```

en tête quand j'ai ouvert le ticket sur les listes #1271, mais en fait même ça ça n'est pas idéal car on duplique plein de fois le "applicable si" alors que l'on voudrait exprimer plus généralement la règle d'applicabilité des taux réduits.

Peut-être 

```yaml
barème:
  multiplicateur:
    valeur: prorata
    applicable si: éligible taux réduit
```

Reply:
> on duplique plein de fois le "applicable si"
Ce n'est pas le applicable si qui est dupliqué, mais bien la première tranche du taux réduit. Dans l'idéal c'est cette dernière qu'il faudrait factoriser.

Mais sur le fond : je ne comprends pas vraiment la logique derrière le choix de cette écriture. Pouquoi pas : 
```yaml
  
    multiplicateur: prorata durée de l'exercice
    variations:
        - si: exercice . début >= 01/01/2022
          alors:
            tranches:
              - taux: 15%
                plafond: plafond taux réduit
              - taux: 25%
        - si: exercice . début >= 01/01/2021
          alors:
            tranches:
              - taux: 15%
                plafond: plafond taux réduit
              - taux: 26.5%
        - si: exercice . début >= 01/01/2020
          alors:
            tranches:
              - taux: 15%
                plafond: plafond taux réduit
              - taux: 28%
        - si: exercice . début >= 01/01/2019
          alors:
            tranches:
              - taux: 15%
                plafond: plafond taux réduit
              - taux: 28%
                plafond: 
                  valeur: 500000 €/an
                  applicable si: taux réduit              
              - taux: 31%
        - si: exercice . début >= 01/01/2018
          alors:
            tranches:
              - taux: 15%
                plafond: plafond taux réduit
              - taux: 28%
                plafond: 
                  valeur: 500000 €/an
                  applicable si: taux réduit
              - taux: 33.3333%
    arrondi: oui

plafond taux réduit: 
  valeur: 38120 €/an 
  applicable si: taux réduit

```
Qui me parait plus clair 
=> Voir le [publi.code studio](https://publi.codes/studio/IS?code=exercice%3A%20oui%0Aexercice%20.%20d%C3%A9but%3A%2001%2F01%2F2018%0Ab%C3%A9n%C3%A9fice%3A%205000%E2%82%AC%2Fan%0Aprorata%20dur%C3%A9e%20de%20l%27exercice%3A%200.4%0Ataux%20r%C3%A9duit%3A%20oui%0AIS%3A%0A%20%20bar%C3%A8me%3A%0A%20%20%20%20assiette%3A%20b%C3%A9n%C3%A9fice%0A%20%20%20%20multiplicateur%3A%20prorata%20dur%C3%A9e%20de%20l%27exercice%0A%20%20%20%20variations%3A%0A%20%20%20%20%20%20%20%20-%20si%3A%20exercice%20.%20d%C3%A9but%20%3E%3D%2001%2F01%2F2022%0A%20%20%20%20%20%20%20%20%20%20alors%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20tranches%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2015%25%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20plafond%3A%20plafond%20taux%20r%C3%A9duit%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2025%25%0A%20%20%20%20%20%20%20%20-%20si%3A%20exercice%20.%20d%C3%A9but%20%3E%3D%2001%2F01%2F2021%0A%20%20%20%20%20%20%20%20%20%20alors%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20tranches%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2015%25%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20plafond%3A%20plafond%20taux%20r%C3%A9duit%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2026.5%25%0A%20%20%20%20%20%20%20%20-%20si%3A%20exercice%20.%20d%C3%A9but%20%3E%3D%2001%2F01%2F2020%0A%20%20%20%20%20%20%20%20%20%20alors%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20tranches%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2015%25%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20plafond%3A%20plafond%20taux%20r%C3%A9duit%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2028%25%0A%20%20%20%20%20%20%20%20-%20si%3A%20exercice%20.%20d%C3%A9but%20%3E%3D%2001%2F01%2F2019%0A%20%20%20%20%20%20%20%20%20%20alors%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20tranches%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2015%25%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20plafond%3A%20plafond%20taux%20r%C3%A9duit%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2028%25%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20plafond%3A%20%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20valeur%3A%20500000%20%E2%82%AC%2Fan%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20applicable%20si%3A%20taux%20r%C3%A9duit%20%20%20%20%20%20%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2031%25%0A%20%20%20%20%20%20%20%20-%20si%3A%20exercice%20.%20d%C3%A9but%20%3E%3D%2001%2F01%2F2018%0A%20%20%20%20%20%20%20%20%20%20alors%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20tranches%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2015%25%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20plafond%3A%20plafond%20taux%20r%C3%A9duit%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2028%25%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20plafond%3A%20%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20valeur%3A%20500000%20%E2%82%AC%2Fan%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20applicable%20si%3A%20taux%20r%C3%A9duit%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20-%20taux%3A%2033.3333%25%0A%20%20%20%20arrondi%3A%20oui%0A%0Aplafond%20taux%20r%C3%A9duit%3A%20%0A%20%20valeur%3A%2038120%20%E2%82%AC%2Fan%20%0A%20%20applicable%20si%3A%20taux%20r%C3%A9duit%0A)


```

Reply:
J'avais du penser que les plafonds étaient revalorisables / modifiables d'une année sur l'autre mais effectivement c'est plus simple comme ça.